typing/ctype.ml

```diff
@@ -4157,7 +4167,7 @@ let rec eqtype rename type_pairs subst env t1 t2 =
               eqtype_list rename type_pairs subst env tl1 tl2
           | (Tconstr (p1, tl1, _), Tconstr (p2, tl2, _))
                 when Path.same p1 p2 ->
-              eqtype_list rename type_pairs subst env tl1 tl2
+              eqtype_list_same_length rename type_pairs subst env tl1 tl2
```

Comment:
(I'm not sure why we feel so confident that the two argument lists have the same length here.)

Reply:
This is actually something I've wondered about when working on the compiler: what invariants exist for `type_expr`s? For example, I *think* we have the invariant that the level of a `type_expr` is always greater than or equal to the level of any type accessible from it. What other invariants do we have? It would be good to document them.

One invariant I know we do not have is that a `Tconstr` satisfies constraints. This is because we sometimes apply a path to fresh variables and then unify in order to satisfy constraints. So we mostly satisfy constraints, but not quite always.

Yet I think, as I've read the compiler, that we do always have it such that the arity of a type constructor is respected in a `Tconstr`. If I've gotten this wrong in this patch, the result is not catastrophic: it would lead to an exception being thrown in an erroneous program. That's a terrible user experience, but it is not unsound.

In any case, I'm content to remove this tiny optimization if others are made nervous by it.

Reply:
I agree that the invariant `"Tconstr"s have consistent arity` is already expected in many parts of the compiler, it seems fine to me to use it there too.